### ISSUE TITLE
Impossible to change input buffer for async boundary after updating akka-stream from 2.6.14 to any new version

### ISSUE THREAD

**Issue by dssysolyatin** [akka/akka#30935]
It is impossible to change input buffer size for async boundary after updating akka-stream from 2.6.14 to 2.6.15.

The following code does not work with akka 2.6.15-2.6.17, but works with 2.6.14

```
package com.test

import akka.actor.ActorSystem
import akka.dispatch.Dispatchers
import akka.stream.scaladsl.{Flow, Source}

import java.util.concurrent.atomic.AtomicInteger
import scala.concurrent.duration.Duration
import scala.concurrent.{Await, Promise}

object TestAsyncInputBuffer extends App {
  implicit val system: ActorSystem = ActorSystem("TestAsyncInputBuffer")

  val amountRecordReadedFromIterator = new AtomicInteger(0)
  val promise = Promise[Unit]

  val flow = Source
    .fromIterator(() => Iterator.continually(1).map { v => amountRecordReadedFromIterator.incrementAndGet(); v })
    .via(Flow[Int].map { _ =>
      Await.result(promise.future, Duration.Inf)
    }.async(Dispatchers.DefaultBlockingDispatcherId, 1))

  flow.run()

  Thread.sleep(200)
  assert(amountRecordReadedFromIterator.get() <= 2, s"number of elements in the buffer ${amountRecordReadedFromIterator.get()} >= 1")

  promise.trySuccess(())
  system.terminate()
}
```

**Comment by johanandren** [comment]
BTW I also suspect a workaround using the `inputBuffer` attribute should work, if there is a bug.

**Comment by dssysolyatin** [comment]
I looked a little bit deeper into the problem and it turned out that `via(flow.async())` creates two async boundaries (BatchingActorInputBoundary): One for moving data from `Source.fromIterator` to flow inside `via` (in this case buffer size = 16) and another for moving data from flow inside `via` to sink (in this case buffer size = 1).

The following small code fix test:
`
Source
    .fromIterator(() => Iterator.continually(1).map { v => amountRecordReadedFromIterator.incrementAndGet(); v })
    .withAttribute(Attributes.inputBuffer(1, 1))
`

The question is why BatchingActorInputBoundary is created for moving data from `Source.fromIterator` to flow inside `via`. In my understanding the flow before via should be executed inside dispatcher that I specified in async. 
And another question is why the attribute name is`inputBuffer` ? Judging by the logic the name should be `outputBuffer`

**Comment by johanandren** [comment]
I can confirm that this is a regression caused by #30195, I'll see if it is possible to fix without having to revert that fix.

It is called `inputBuffer` because it determines the buffer in front of an async stream "island", which by default, batches are fetched into. While this should work as you expect, I should add that tuning it down to 1 to get that level of flow control anywhere else than in a test is a bad idea, the default windowed batching is there for a reason (see docs https://doc.akka.io/docs/akka/current/stream/stream-rate.html#buffers-for-asynchronous-operators).

**Comment by dssysolyatin** [comment]
@johanandren I know for what it was introduced, but in some cases, I want to have 1 element per worker in the buffer. 

For example, there is a flow:
```
balancer -> workers -> merge
```
If the worker buffer is 16 then it can lead to a situation where only one worker works and another do nothing.


There is another thing that I found about async boundary buffer. The following function will work much faster than inputBuffer = 2048. It is mostly because dispatcher has `throughput` = 100 by default. InputBuffer can have only 100 elements in this case, it can be fixed by creating a separate dispatcher and increasing throughput to 2048, but it is not good to create a separate dispatcher for every async boundary.

```
def sourceAsyncBoundary[O: ClassTag, M](maxBatchSize: Int, inputBufferSize: Int = 16, dispatcherName: String = "akka.actor.default-dispatcher")(source: Source[O, M]): Source[O, M] = {
    source
      .batch(maxBatchSize, el => Vector(el))((acc, el) => acc :+ el)
      .async(dispatcherName, inputBufferSize).mapConcat(identity)
  } 
  ```

**Comment by jrudolph** [comment]
> The question is why is the attribute name`inputBuffer`? Judging by the logic the name should be `outputBuffer`

Exactly, #30195 muddied the waters in that regard and we should revert it. Implementation-wise it always has been an InputBuffer because the buffer is owned by the incoming side of a stream connection. With #30195 it is still an `inputBuffer` but one whose settings is now being overridden by the sending side.
